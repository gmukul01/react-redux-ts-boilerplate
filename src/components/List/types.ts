import { HTMLProps, WithThemeProp } from '../../utils/types';

export interface Props extends HTMLProps<HTMLUListElement>, WithThemeProp {
    variant?: 'horizontal' | 'vertical';
}
