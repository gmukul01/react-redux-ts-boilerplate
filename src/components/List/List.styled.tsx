/* stylelint-disable  no-descending-specificity*/
import { defaultTheme, SpacingTheme } from '@theme';
import { css, styled } from '@utils/styled';
import { Props } from './types';

const horizontal = (spacings: SpacingTheme) => css`
    li {
        * {
            display: inline;
            margin: 0px ${spacings.XS};
        }
        &:last-child {
            * {
                margin-right: ${spacings.S};
            }
        }
        &:first-child {
            * {
                margin-left: ${spacings.S};
            }
        }
    }
`;

const vertical = (spacings: SpacingTheme) => css`
    li {
        * {
            display: block;
            margin: ${spacings.XS} 0px;
        }
        &:last-child {
            * {
                margin-bottom: ${spacings.S};
            }
        }
        &:first-child {
            * {
                margin-top: ${spacings.S};
            }
        }
    }
`;

const ListStyled = styled('ul').attrs(({ theme: { spacing: spacings } }) => ({ spacings }))<Props>`
    padding: 0;
    margin: 0;

    ${({ spacings, variant }) => variant === 'horizontal' && horizontal(spacings)};
    ${({ spacings, variant }) => variant === 'vertical' && vertical(spacings)};
`;

ListStyled.defaultProps = {
    theme: defaultTheme
};

export { ListStyled };
