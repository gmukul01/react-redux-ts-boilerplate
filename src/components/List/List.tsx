import React from 'react';
import { WithStyle } from '@utils/types';
import { ListStyled } from './List.styled';
import { Props } from './types';

const List: React.FunctionComponent<Props> & WithStyle = props => {
    return (
        <ListStyled {...props}>
            {React.Children.map(props.children, (c, i) => {
                return <li key={i}>{c}</li>;
            })}
        </ListStyled>
    );
};

List.displayName = 'List';
List.Style = ListStyled;
List.defaultProps = {
    variant: 'vertical'
};

export default List;
