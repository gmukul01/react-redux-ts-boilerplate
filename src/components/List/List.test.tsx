import 'jest-styled-components';
import React from 'react';
import { create } from 'react-test-renderer';

import List from './List';

describe('List component', () => {
    it('should render horizontal list', () => {
        const list = create(
            <List variant="horizontal">
                <p>item 1</p>
                <p>item 2</p>
            </List>
        );
        expect(list).toMatchSnapshot();
    });

    it('should render vertical list', () => {
        const list = create(
            <List variant="vertical">
                <p>item 1</p>
                <p>item 2</p>
            </List>
        );
        expect(list).toMatchSnapshot();
    });
});
