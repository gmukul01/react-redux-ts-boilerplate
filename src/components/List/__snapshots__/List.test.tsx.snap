// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render horizontal list 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  list-style-type: none;
}

.c0 li * {
  display: inline;
  margin: 0px 4px;
}

.c0 li:last-child * {
  margin-right: 8px;
}

.c0 li:first-child * {
  margin-left: 8px;
}

<div>
  <ul
    class="c0"
  >
    <li>
      <p>
        item 1
      </p>
    </li>
    <li>
      <p>
        item 2
      </p>
    </li>
  </ul>
</div>
`;

exports[`List component should render vertical list 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  list-style-type: none;
}

.c0 li * {
  display: block;
  margin: 4px 0px;
}

.c0 li:last-child * {
  margin-bottom: 8px;
}

.c0 li:first-child * {
  margin-top: 8px;
}

<div>
  <ul
    class="c0"
  >
    <li>
      <p>
        item 1
      </p>
    </li>
    <li>
      <p>
        item 2
      </p>
    </li>
  </ul>
</div>
`;
